spring-integration-hazelcast/src/main/java/org/springframework/integration/hazelcast/metadata/HazelcastMetadataStore.java

```diff
@@ -0,0 +1,131 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.hazelcast.metadata;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import com.hazelcast.core.EntryEvent;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.IMap;
+import com.hazelcast.map.listener.EntryAddedListener;
+import com.hazelcast.map.listener.EntryRemovedListener;
+import com.hazelcast.map.listener.EntryUpdatedListener;
+
+import org.springframework.beans.factory.InitializingBean;
+import org.springframework.integration.metadata.ListenableMetadataStore;
+import org.springframework.integration.metadata.MetadataStoreListener;
+import org.springframework.util.Assert;
+
+/**
+ * @author Vinicius Carvalho
+ */
+public class HazelcastMetadataStore implements ListenableMetadataStore, InitializingBean{
+
+	private static final String METADATA_STORE_MAP_NAME = "SPRING_INTEGRATION_METADATA_STORE";
+
+	private final IMap<String, String> map;
+
+	private final List<MetadataStoreListener> listeners = new CopyOnWriteArrayList<MetadataStoreListener>();
+
+	public HazelcastMetadataStore(HazelcastInstance hazelcastInstance){
+		Assert.notNull(hazelcastInstance, "Hazelcast instance can't be null");
+		this.map = hazelcastInstance.getMap(METADATA_STORE_MAP_NAME);
+	}
+
+	public HazelcastMetadataStore(IMap<String, String> map) {
+		Assert.notNull(map, "IMap reference can not be null");
+		this.map = map;
+	}
+
+	@Override
+	public String putIfAbsent(String key, String value) {
+		Assert.notNull(key, "'key' must not be null.");
+		Assert.notNull(value, "'value' must not be null.");
+		return this.map.putIfAbsent(key,value);
+	}
+
+	@Override
+	public boolean replace(String key, String oldValue, String newValue) {
+		Assert.notNull(key, "'key' must not be null.");
+		Assert.notNull(oldValue, "'oldValue' must not be null.");
+		Assert.notNull(newValue, "'newValue' must not be null.");
+		return this.map.replace(key,oldValue,newValue);
+	}
+
+	@Override
+	public void put(String key, String value) {
+		Assert.notNull(key, "'key' must not be null.");
+		Assert.notNull(value, "'value' must not be null.");
+		this.map.put(key,value);
+	}
+
+	@Override
+	public String get(String key) {
+		Assert.notNull(key, "'key' must not be null.");
+		return this.map.get(key);
+	}
+
+	@Override
+	public String remove(String key) {
+		Assert.notNull(key, "'key' must not be null.");
+		return this.map.remove(key);
+	}
+
+	@Override
+	public void addListener(MetadataStoreListener callback) {
+		this.listeners.add(callback);
+	}
+
+	@Override
+	public void removeListener(MetadataStoreListener callback) {
+		this.listeners.remove(callback);
+	}
+
+	@Override
+	public void afterPropertiesSet() throws Exception {
+		this.map.addEntryListener(new MapListener(),true);
+	}
+
+	final class MapListener implements  EntryAddedListener<String,String>,
```

Comment:
>I can't recall now the source of this motivation

With a private inner class, if the CTOR is private, the compiler has to make a synthetic, package-protected, ctor that takes an extra dummy argument and delegates to the actual ctor. Making the ctor package protected avoids this extra call in the call stack.

We generally prefer static inner classes (unless there are a lot of dependencies on outer class fields) because they are more efficient. An extra argument (this) is added to every method call.